File: backend/app/schemas/advanced_editor.py
"""Pydantic schemas for Advanced Editor."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from app.models import (
    AssetType,
    BlendMode,
    ItemType,
    ProjectStatus,
    SourceType,
    TrackType,
    TransitionType,
)


# Project Schemas
class ProjectCreate(BaseModel):
    """Schema for creating a project."""

    name: str = Field(..., max_length=255)
    description: str | None = None
    video_id: UUID | None = None
    width: int = 1920
    height: int = 1080
    frame_rate: float = 30.0
    duration_seconds: float = 60.0
    background_color: str = "#000000"


class ProjectUpdate(BaseModel):
    """Schema for updating a project."""

    name: str | None = None
    description: str | None = None
    width: int | None = None
    height: int | None = None
    frame_rate: float | None = None
    duration_seconds: float | None = None
    background_color: str | None = None
    canvas_settings: dict | None = None
    export_settings: dict | None = None


class ProjectRead(BaseModel):
    """Schema for reading a project."""

    id: UUID
    user_id: UUID
    name: str
    description: str | None
    video_id: UUID | None
    width: int
    height: int
    frame_rate: float
    duration_seconds: float
    background_color: str
    canvas_settings: dict
    export_settings: dict
    thumbnail_url: str | None
    status: ProjectStatus
    last_rendered_at: datetime | None
    render_output_url: str | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class RenderConfig(BaseModel):
    """Schema for render configuration."""

    quality: str = Field(default="high", pattern="^(low|medium|high|max)$")
    format: str = Field(default="mp4", pattern="^(mp4|mov|webm)$")
    resolution: dict | None = None


class RenderProgress(BaseModel):
    """Schema for render progress."""

    progress: int = Field(..., ge=0, le=100)
    stage: str
    status: str
    estimated_time_remaining: float | None = None
    error: str | None = None


class ValidationResult(BaseModel):
    """Schema for project validation result."""

    valid: bool
    errors: list[str]
    warnings: list[str]


# Track Schemas
class TrackCreate(BaseModel):
    """Schema for creating a track."""

    track_type: TrackType
    name: str = Field(..., max_length=255)
    z_index: int | None = None
    volume: float = Field(default=1.0, ge=0.0, le=1.0)
    opacity: float = Field(default=1.0, ge=0.0, le=1.0)


class TrackUpdate(BaseModel):
    """Schema for updating a track."""

    name: str | None = None
    z_index: int | None = None
    is_locked: bool | None = None
    is_visible: bool | None = None
    is_muted: bool | None = None
    volume: float | None = Field(default=None, ge=0.0, le=1.0)
    opacity: float | None = Field(default=None, ge=0.0, le=1.0)
    blend_mode: BlendMode | None = None
    track_order: int | None = None


class TrackRead(BaseModel):
    """Schema for reading a track."""

    id: UUID
    project_id: UUID
    track_type: TrackType
    name: str
    z_index: int
    is_locked: bool
    is_visible: bool
    is_muted: bool
    volume: float
    opacity: float
    blend_mode: BlendMode
    track_order: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# TrackItem Schemas
class TrackItemCreate(BaseModel):
    """Schema for creating a track item."""

    item_type: ItemType
    source_type: SourceType
    source_id: UUID | None = None
    source_url: str | None = None
    start_time: float = Field(..., ge=0.0)
    end_time: float = Field(..., ge=0.0)
    trim_start: float = Field(default=0.0, ge=0.0)
    trim_end: float = Field(default=0.0, ge=0.0)
    position_x: float = Field(default=0.0, ge=0.0, le=1.0)
    position_y: float = Field(default=0.0, ge=0.0, le=1.0)
    scale_x: float = Field(default=1.0, ge=0.0)
    scale_y: float = Field(default=1.0, ge=0.0)
    rotation: float = Field(default=0.0)
    crop_settings: dict | None = None
    transform_settings: dict | None = None
    text_content: str | None = None
    text_style: dict | None = None
    effects: list | None = None
    transition_in_id: UUID | None = None
    transition_out_id: UUID | None = None


class TrackItemUpdate(BaseModel):
    """Schema for updating a track item."""

    item_type: ItemType | None = None
    source_type: SourceType | None = None
    source_id: UUID | None = None
    source_url: str | None = None
    start_time: float | None = Field(default=None, ge=0.0)
    end_time: float | None = Field(default=None, ge=0.0)
    trim_start: float | None = Field(default=None, ge=0.0)
    trim_end: float | None = Field(default=None, ge=0.0)
    position_x: float | None = Field(default=None, ge=0.0, le=1.0)
    position_y: float | None = Field(default=None, ge=0.0, le=1.0)
    scale_x: float | None = Field(default=None, ge=0.0)
    scale_y: float | None = Field(default=None, ge=0.0)
    rotation: float | None = None
    crop_settings: dict | None = None
    transform_settings: dict | None = None
    text_content: str | None = None
    text_style: dict | None = None
    effects: list | None = None
    transition_in_id: UUID | None = None
    transition_out_id: UUID | None = None


class TrackItemRead(BaseModel):
    """Schema for reading a track item."""

    id: UUID
    track_id: UUID
    item_type: ItemType
    source_type: SourceType
    source_id: UUID | None
    source_url: str | None
    start_time: float
    end_time: float
    duration: float
    trim_start: float
    trim_end: float
    position_x: float
    position_y: float
    scale_x: float
    scale_y: float
    rotation: float
    crop_settings: dict | None
    transform_settings: dict | None
    text_content: str | None
    text_style: dict | None
    effects: list | None
    transition_in_id: UUID | None
    transition_out_id: UUID | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SplitItemRequest(BaseModel):
    """Schema for splitting a track item."""

    split_time: float = Field(..., gt=0.0)


class SplitItemResponse(BaseModel):
    """Schema for split item response."""

    item1: TrackItemRead
    item2: TrackItemRead


# Asset Schemas
class AssetCreate(BaseModel):
    """Schema for creating an asset (used with multipart form data)."""

    asset_type: AssetType
    name: str = Field(..., max_length=255)
    tags: list[str] = Field(default_factory=list)


class AssetUpdate(BaseModel):
    """Schema for updating an asset."""

    name: str | None = Field(default=None, max_length=255)
    tags: list[str] | None = None


class AssetRead(BaseModel):
    """Schema for reading an asset."""

    id: UUID
    user_id: UUID
    asset_type: AssetType
    name: str
    file_url: str
    file_size: int | None
    mime_type: str | None
    width: int | None
    height: int | None
    duration_seconds: float | None
    metadata: dict
    thumbnail_url: str | None
    is_public: bool
    tags: list[str]
    usage_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Transition Schemas
class TransitionRead(BaseModel):
    """Schema for reading a transition."""

    id: UUID
    name: str
    transition_type: TransitionType
    direction: str | None
    default_duration: float
    parameters: dict
    preview_url: str | None
    is_builtin: bool
    is_public: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TransitionCreate(BaseModel):
    """Schema for creating a custom transition."""

    name: str = Field(..., max_length=255)
    transition_type: TransitionType
    direction: str | None = None
    default_duration: float = Field(default=0.5, gt=0.0)
    parameters: dict = Field(default_factory=dict)

File: frontend/src/types/advancedEditor.ts
// TypeScript type definitions for Advanced Editor feature

// ==================== Enums ====================

export type ProjectStatus = 'draft' | 'rendering' | 'completed' | 'error';

export type TrackType = 'video' | 'audio' | 'image' | 'text' | 'overlay';

export type ItemType = 'video_clip' | 'audio_clip' | 'image' | 'text' | 'shape';

export type SourceType = 'video' | 'asset' | 'text' | 'generated';

export type AssetType = 'image' | 'audio' | 'font' | 'graphic' | 'template';

export type TransitionType = 'fade' | 'dissolve' | 'slide' | 'wipe' | 'zoom' | 'blur' | 'custom';

export type TransitionDirection = 'in' | 'out';

export type EffectType = 'filter' | 'color' | 'blur' | 'sharpen' | 'distort' | 'custom';

export type BlendMode = 'normal' | 'multiply' | 'screen' | 'overlay';

export type RenderQuality = 'low' | 'medium' | 'high' | 'max';

export type RenderFormat = 'mp4' | 'mov' | 'webm';

export type RenderStatus = 'queued' | 'processing' | 'completed' | 'error';

export type TextAlignment = 'left' | 'center' | 'right';

// ==================== Interfaces ====================

// Project
export interface Project {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  video_id?: string;
  width: number;
  height: number;
  frame_rate: number;
  duration_seconds: number;
  background_color: string;
  canvas_settings?: Record<string, any>;
  export_settings?: Record<string, any>;
  thumbnail_url?: string;
  status: ProjectStatus;
  last_rendered_at?: string;
  render_output_url?: string;
  created_at: string;
  updated_at: string;
  tracks?: Track[];
}

// Track
export interface Track {
  id: string;
  project_id: string;
  track_type: TrackType;
  name: string;
  z_index: number;
  is_locked: boolean;
  is_visible: boolean;
  is_muted: boolean;
  volume: number;
  opacity: number;
  blend_mode: BlendMode;
  track_order: number;
  created_at: string;
  updated_at: string;
  items?: TrackItem[];
}

// TrackItem
export interface TrackItem {
  id: string;
  track_id: string;
  item_type: ItemType;
  source_type: SourceType;
  source_id?: string;
  source_url?: string;
  start_time: number;
  end_time: number;
  duration: number;
  trim_start: number;
  trim_end: number;
  position_x: number;
  position_y: number;
  scale_x: number;
  scale_y: number;
  rotation: number;
  crop_settings?: CropSettings;
  transform_settings?: TransformSettings;
  text_content?: string;
  text_style?: TextStyle;
  effects?: any[];
  transition_in?: string;
  transition_out?: string;
  created_at: string;
  updated_at: string;
}

// Asset
export interface Asset {
  id: string;
  user_id: string;
  asset_type: AssetType;
  name: string;
  file_url: string;
  file_size: number;
  mime_type: string;
  width?: number;
  height?: number;
  duration_seconds?: number;
  metadata?: Record<string, any>;
  thumbnail_url?: string;
  is_public: boolean;
  tags: string[];
  usage_count: number;
  created_at: string;
  updated_at: string;
}

// Transition
export interface Transition {
  id: string;
  name: string;
  transition_type: TransitionType;
  direction?: TransitionDirection;
  default_duration: number;
  parameters?: Record<string, any>;
  preview_url?: string;
  is_builtin: boolean;
  is_public: boolean;
  created_at: string;
  updated_at: string;
}

// CompositionEffect
export interface CompositionEffect {
  id: string;
  name: string;
  effect_type: EffectType;
  parameters: Record<string, any>;
  ffmpeg_filter?: string;
  is_builtin: boolean;
  preview_url?: string;
  created_at: string;
}

// ==================== Configuration Interfaces ====================

// ProjectConfig
export interface ProjectConfig {
  name: string;
  description?: string;
  video_id?: string;
  width: number;
  height: number;
  frame_rate: number;
  duration: number;
}

// TrackConfig
export interface TrackConfig {
  track_type: TrackType;
  name: string;
  z_index?: number;
  volume?: number;
  opacity?: number;
}

// TrackItemConfig
export interface TrackItemConfig {
  item_type: ItemType;
  source_type: SourceType;
  source_id?: string;
  source_url?: string;
  start_time: number;
  end_time: number;
  position_x?: number;
  position_y?: number;
  scale_x?: number;
  scale_y?: number;
  rotation?: number;
  text_content?: string;
  text_style?: TextStyle;
  transition_in?: string;
  transition_out?: string;
}

// RenderConfig
export interface RenderConfig {
  quality: RenderQuality;
  format: RenderFormat;
  resolution?: {
    width: number;
    height: number;
  };
  codec?: string;
  bitrate?: number;
  audio_codec?: string;
  audio_bitrate?: number;
}

// RenderProgress
export interface RenderProgress {
  progress: number; // 0-100
  stage: string;
  status: RenderStatus;
  estimated_remaining: number; // seconds
}

// ValidationResult
export interface ValidationResult {
  valid: boolean;
  errors: string[];
  warnings: string[];
}

// ==================== Nested Property Interfaces ====================

// TextStyle
export interface TextStyle {
  font_family: string;
  font_size: number;
  color: string;
  alignment: TextAlignment;
  bold: boolean;
  italic: boolean;
  underline: boolean;
  shadow: boolean;
  outline: boolean;
}

// CropSettings
export interface CropSettings {
  top: number; // pixels
  bottom: number; // pixels
  left: number; // pixels
  right: number; // pixels
}

// TransformSettings
export interface TransformSettings {
  skew_x?: number;
  skew_y?: number;
  flip_horizontal?: boolean;
  flip_vertical?: boolean;
}

// ==================== Response Interfaces ====================

// API response wrappers
export interface ProjectsResponse {
  projects: Project[];
  total: number;
}

export interface AssetsResponse {
  assets: Asset[];
  total: number;
}

export interface TransitionsResponse {
  transitions: Transition[];
}

export interface RenderJobResponse {
  job_id: string;
  estimated_time: number;
}

export interface SplitItemResponse {
  item1: TrackItem;
  item2: TrackItem;
}

export interface DeleteResponse {
  message: string;
}

// ==================== Resolution Presets ====================

export interface ResolutionPreset {
  name: string;
  width: number;
  height: number;
  description: string;
}

export const RESOLUTION_PRESETS: ResolutionPreset[] = [
  { name: '1080p (Full HD)', width: 1920, height: 1080, description: 'Standard HD video' },
  { name: '720p (HD)', width: 1280, height: 720, description: 'HD video, smaller file size' },
  { name: '4K (Ultra HD)', width: 3840, height: 2160, description: 'Ultra high definition' },
  { name: 'Vertical (9:16)', width: 1080, height: 1920, description: 'For Instagram Stories, TikTok' },
  { name: 'Square (1:1)', width: 1080, height: 1080, description: 'For Instagram Feed' },
  { name: 'Custom', width: 0, height: 0, description: 'Custom resolution' },
];

// ==================== Quality Presets ====================

export interface QualityPreset {
  name: RenderQuality;
  label: string;
  resolution: string;
  bitrate: string;
  description: string;
}

export const QUALITY_PRESETS: QualityPreset[] = [
  { name: 'low', label: 'Low', resolution: '720p', bitrate: '2 Mbps', description: 'Fast export, smaller file size' },
  { name: 'medium', label: 'Medium', resolution: '1080p', bitrate: '5 Mbps', description: 'Balanced quality and size' },
  { name: 'high', label: 'High', resolution: '1080p', bitrate: '10 Mbps', description: 'High quality, larger file' },
  { name: 'max', label: 'Maximum', resolution: '4K', bitrate: '20 Mbps', description: 'Best quality, very large file' },
];

// ==================== Utility Types ====================

// For partial updates
export type ProjectUpdate = Partial<Omit<Project, 'id' | 'user_id' | 'created_at' | 'updated_at'>>;
export type TrackUpdate = Partial<Omit<Track, 'id' | 'project_id' | 'created_at' | 'updated_at'>>;
export type TrackItemUpdate = Partial<Omit<TrackItem, 'id' | 'track_id' | 'created_at' | 'updated_at'>>;
export type AssetUpdate = Partial<Pick<Asset, 'name' | 'tags'>>;

// For query parameters
export interface ProjectsQueryParams {
  status?: ProjectStatus;
  limit?: number;
  offset?: number;
  sort?: string;
}

export interface AssetsQueryParams {
  asset_type?: AssetType;
  tags?: string[];
  limit?: number;
  offset?: number;
  search?: string;
}

export interface TransitionsQueryParams {
  transition_type?: TransitionType;
  is_public?: boolean;
}
